### PR TITLE
Fix/refactor dynamo onnxrt backend

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #93878
* #93871
* #93870
* #93863
* #93822
* #93821
* __->__ #93818

Fixes https://github.com/pytorch/pytorch/issues/90352

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/backends/common.py [hunk 1]
```diff
@@ -70,3 +73,37 @@ def mem_efficient_fusion_kwargs(use_decomps):
...
+        if not isinstance(x, FakeTensor):
+            return x
+        y = torch.empty_strided(
+            x.size(),
+            x.stride(),
+            dtype=x.dtype,
+            device=x.device,
+            requires_grad=x.requires_grad,
+        )
+        y.zero_()
```
Comment (voznesenskym) [inline torch/_dynamo/backends/common.py:94]: layout?

Comment (jansel) [reply torch/_dynamo/backends/common.py:94]: zero_ doesn't change the layout set above.

File: torch/_dynamo/backends/common.py [hunk 2]
```diff
@@ -70,3 +73,37 @@ def mem_efficient_fusion_kwargs(use_decomps):
...
+    @functools.wraps(fn)
+    def wrapper(model, inputs, **kwargs):
+        with _disable_current_modes():
+            inputs = list(map(defake, inputs))
+            return fn(model, inputs, **kwargs)
+
+    return wrapper
+
+
+def device_from_inputs(example_inputs) -> torch.device:
```
Comment (voznesenskym) [inline torch/_dynamo/backends/common.py:106]: nit: move to utils? 

Also - would we ever have multiple devices?

**Review by voznesenskym — APPROVED** [review]

File: torch/_dynamo/backends/common.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/backends/common.py:94]: I don't see you passing layout to empty_strided though?

**Pull request closed without merging**